Fade animation for newly added Markdown text chunks

mg3994 opened the pull request:
#### Must Modify 
- Durations as per your need

### Changes Made

0. **Use of Keys**: Makes the animation smoother by ensuring unique widget identity, leading to more efficient animation transitions.
   
1. **Used a Single `MarkdownBody` Widget**: This ensures that all markdown text is rendered smoothly in one widget, avoiding redundancy.

2. **Animated Only the New Chunk of Text**: The `AnimatedOpacity` widget is used specifically for the newly added text chunk, while `AnimatedSwitcher` controls how the new text enters. 

3. **Text Rendering**: The text is rendered using the `MarkdownBody` widget. Animation is applied only to the newly added text chunk when the floating action button is pressed.

4. **Scroll to Bottom**: The scroll automatically moves to the bottom after new text is added (you can modify this behavior as needed).

5. **Easy to Customize**: The implementation features less code with simple logic, making it easy to adjust as per your requirements.

### Layout Modifications

- The `layoutBuilder` has been modified and can be customized as needed:

```dart
layoutBuilder: (currentChild, previousChildren) {  // Customize as per your need
  return Stack(
    children: [
      // Display previous text statically
      ...previousChildren,

      // Apply fade-in animation to the new chunk of text
      if (currentChild != null)
        AnimatedOpacity(
          opacity: 1.0,
          duration: const Duration(milliseconds: 800), // Modify duration as needed
          child: currentChild,
        ),
    ],
  );
},
```
Feel free to let me know if you need any modifications or additional customizations. I’m open to making adjustments to better fit your needs, whether it's tweaking animation durations, changing transition effects, or any other enhancements you have in mind.

Looking forward to your feedback!


https://github.com/user-attachments/assets/3e72d63f-9421-4c4d-b865-3faf7ed267d0

mg3994 committed:
making code smaller removing useless stuffs,

TODO: is in pending, must fix

mg3994 committed:
Update main.dart

mg3994 committed:
Update main.dart

FilledStacks commented:
This the closest to what we need.

Thank you for taking the time to actually understand the requirement. 

The issue I have here is that the already rendered text pulses slightly, which is an undesired after effect. 

If you can fix that then this solution will work well. 

Looking forward to the next PR

mg3994 committed:
Make it Simple to Understand not too hard

mg3994 commented:
The Code given below introduces a feature that enables the markdown text to gradually fade in as it is streamed or added to the screen. The animation ensures that each new chunk of text is rendered using the markdown renderer with a smooth fade-in effect. (I also Modified the chunks markdown List , so that you can test it easily)
#### Do change `Curves` that fits it best (for that kind of animation)
> Copy Paste the Below code, I made it simple with Opacity and removed other extra stuffs
```dart
import 'package:flutter/material.dart';
import 'package:flutter_markdown_selectionarea/flutter_markdown_selectionarea.dart';

class FadeRevealMarkdownDifference extends StatefulWidget {
  const FadeRevealMarkdownDifference({super.key});

  @override
  State<FadeRevealMarkdownDifference> createState() =>
      _FadeRevealMarkdownDifferenceState();
}

class _FadeRevealMarkdownDifferenceState
    extends State<FadeRevealMarkdownDifference>
    with SingleTickerProviderStateMixin {
  final List<String> markdownVersions = [
    "## Problem\nThis is the main issue with what",
    "## Problem\nThis is the main issue with what we're trying to solve.",
    "## Problem\nThis is the main issue with what we're trying to solve. The current solution is insufficient for handling edge cases.",
    "## Problem\nThis is the main issue with what we're trying to solve. The current solution is insufficient for handling edge cases, and it leads to frequent breakdowns in production.",
    "## Problem\nThis is the main issue with what we're trying to solve. The current solution is insufficient for handling edge cases, and it leads to frequent breakdowns in production. As a result, customer satisfaction has dropped significantly.",
    "## Problem\nThis is the main issue with what we're trying to solve. The current solution is insufficient for handling edge cases, and it leads to frequent breakdowns in production. As a result, customer satisfaction has dropped significantly. We need a more robust solution.",
    "## Problem\nThis is the main issue with what we're trying to solve. The current solution is insufficient for handling edge cases, and it leads to frequent breakdowns in production. As a result, customer satisfaction has dropped significantly. We need a more robust solution that can handle these edge cases effectively.",
    "## Problem\nThis is the main issue with what we're trying to solve. The current solution is insufficient for handling edge cases, and it leads to frequent breakdowns in production. As a result, customer satisfaction has dropped significantly. We need a more robust solution that can handle these edge cases effectively. This will require a complete overhaul of the system.",
    "## Problem\nThis is the main issue with what we're trying to solve. The current solution is insufficient for handling edge cases, and it leads to frequent breakdowns in production. As a result, customer satisfaction has dropped significantly. We need a more robust solution that can handle these edge cases effectively. This will require a complete overhaul of the system. The new system should be designed with scalability and reliability in mind.",
    "## Problem\nThis is the main issue with what we're trying to solve. The current solution is insufficient for handling edge cases, and it leads to frequent breakdowns in production. As a result, customer satisfaction has dropped significantly. We need a more robust solution that can handle these edge cases effectively. This will require a complete overhaul of the system. The new system should be designed with scalability and reliability in mind. Additionally, it should be user-friendly and easy to maintain.",
    "## Problem\nThis is the main issue with what we're trying to solve. The current solution is insufficient for handling edge cases, and it leads to frequent breakdowns in production. As a result, customer satisfaction has dropped significantly. We need a more robust solution that can handle these edge cases effectively. This will require a complete overhaul of the system. The new system should be designed with scalability and reliability in mind. Additionally, it should be user-friendly and easy to maintain. The development team should also consider implementing automated testing to ensure the system's stability.",
    "## Problem\nThis is the main issue with what we're trying to solve. The current solution is insufficient for handling edge cases, and it leads to frequent breakdowns in production. As a result, customer satisfaction has dropped significantly. We need a more robust solution that can handle these edge cases effectively. This will require a complete overhaul of the system. The new system should be designed with scalability and reliability in mind. Additionally, it should be user-friendly and easy to maintain. The development team should also consider implementing automated testing to ensure the system's stability. Regular user feedback should be collected to continuously improve the system.",
    "## Problem\nThis is the main issue with what we're trying to solve. The current solution is insufficient for handling edge cases, and it leads to frequent breakdowns in production. As a result, customer satisfaction has dropped significantly. We need a more robust solution that can handle these edge cases effectively. This will require a complete overhaul of the system. The new system should be designed with scalability and reliability in mind. Additionally, it should be user-friendly and easy to maintain. The development team should also consider implementing automated testing to ensure the system's stability. Regular user feedback should be collected to continuously improve the system. The project should be managed using agile methodologies to ensure flexibility and adaptability.",
    "## Problem\nThis is the main issue with what we're trying to solve. The current solution is insufficient for handling edge cases, and it leads to frequent breakdowns in production. As a result, customer satisfaction has dropped significantly. We need a more robust solution that can handle these edge cases effectively. This will require a complete overhaul of the system. The new system should be designed with scalability and reliability in mind. Additionally, it should be user-friendly and easy to maintain. The development team should also consider implementing automated testing to ensure the system's stability. Regular user feedback should be collected to continuously improve the system. The project should be managed using agile methodologies to ensure flexibility and adaptability. The team should also focus on documentation and knowledge sharing to ensure smooth onboarding of new team members.",
    "## Problem\nThis is the main issue with what we're trying to solve. The current solution is insufficient for handling edge cases, and it leads to frequent breakdowns in production. As a result, customer satisfaction has dropped significantly. We need a more robust solution that can handle these edge cases effectively. This will require a complete overhaul of the system. The new system should be designed with scalability and reliability in mind. Additionally, it should be user-friendly and easy to maintain. The development team should also consider implementing automated testing to ensure the system's stability. Regular user feedback should be collected to continuously improve the system. The project should be managed using agile methodologies to ensure flexibility and adaptability. The team should also focus on documentation and knowledge sharing to ensure smooth onboarding of new team members. The system should be regularly audited for security vulnerabilities to ensure data protection.",
    "## Problem\nThis is the main issue with what we're trying to solve. The current solution is insufficient for handling edge cases, and it leads to frequent breakdowns in production. As a result, customer satisfaction has dropped significantly. We need a more robust solution that can handle these edge cases effectively. This will require a complete overhaul of the system. The new system should be designed with scalability and reliability in mind. Additionally, it should be user-friendly and easy to maintain. The development team should also consider implementing automated testing to ensure the system's stability. Regular user feedback should be collected to continuously improve the system. The project should be managed using agile methodologies to ensure flexibility and adaptability. The team should also focus on documentation and knowledge sharing to ensure smooth onboarding of new team members. The system should be regularly audited for security vulnerabilities to ensure data protection. The system should also be designed to comply with relevant regulations and standards.",
    "## Problem\nThis is the main issue with what we're trying to solve. The current solution is insufficient for handling edge cases, and it leads to frequent breakdowns in production. As a result, customer satisfaction has dropped significantly. We need a more robust solution that can handle these edge cases effectively. This will require a complete overhaul of the system. The new system should be designed with scalability and reliability in mind. Additionally, it should be user-friendly and easy to maintain. The development team should also consider implementing automated testing to ensure the system's stability. Regular user feedback should be collected to continuously improve the system. The project should be managed using agile methodologies to ensure flexibility and adaptability. The team should also focus on documentation and knowledge sharing to ensure smooth onboarding of new team members. The system should be regularly audited for security vulnerabilities to ensure data protection. The system should also be designed to comply with relevant regulations and standards. The system should be regularly updated to incorporate the latest technologies and best practices.",
  ];

  late AnimationController _controller;
  late Animation<double> _fadeAnimation;

  String previousText = "";
  String currentText = "";
  String newText = "";

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
      duration: const Duration(seconds: 1), // Duration of the fade effect
      vsync: this,
    );

    // Initialize text and animation for the first version
    _updateText(0);
  }

  void _updateText(int versionIndex) {
    if (versionIndex < 0 || versionIndex >= markdownVersions.length) return;

    setState(() {
      previousText = versionIndex > 0 ? markdownVersions[versionIndex - 1] : "";
      currentText = markdownVersions[versionIndex];
      newText = _findNewText(previousText, currentText);

      // Reinitialize animation for the fade effect
      _fadeAnimation = Tween<double>(
        begin: 0.0,
        end: 1.0,
      ).animate(
        CurvedAnimation(
          parent: _controller,
          curve: Curves.easeInOut,
        ),
      );

      // Reset and start the animation
      _controller.reset();
      _controller.forward();
    });
  }

  // Compare previous and current text to return the newly added portion
  String _findNewText(String oldText, String newText) {
    if (oldText == newText) return "";
    return newText.substring(oldText.length); // Get new text after old text
  }

  @override
  void dispose() {
    _controller.dispose(); // Dispose of the controller to free resources
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Fade Reveal Markdown Difference Effect"),
      ),
      body: Padding(
        padding: const EdgeInsets.all(16.0),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Expanded(
              child: AnimatedBuilder(
                animation: _fadeAnimation,
                builder: (context, child) {
                  // Combine the old static markdown and the animated new markdown
                  String fullMarkdown = previousText + newText;

                  return Stack(
                    children: [
                      Opacity(
                        opacity: 1 - _fadeAnimation.value,
                        child: MarkdownBody(
                          data: previousText,
                          styleSheet: MarkdownStyleSheet(
                            h2: const TextStyle(
                                fontSize: 24, fontWeight: FontWeight.bold),
                            p: const TextStyle(
                                fontSize: 18, color: Colors.black87),
                          ),
                        ),
                      ),
                      Opacity(
                        opacity: _fadeAnimation.value,
                        child: MarkdownBody(
                          data: fullMarkdown,
                          styleSheet: MarkdownStyleSheet(
                            h2: const TextStyle(
                                fontSize: 24, fontWeight: FontWeight.bold),
                            p: const TextStyle(
                                fontSize: 18, color: Colors.black87),
                          ),
                        ),
                      ),
                    ],
                  );
                },
              ),
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        child: const Icon(Icons.add),
        onPressed: () {
          // Move to the next version and restart the animation
          int nextIndex = (markdownVersions.indexOf(currentText) + 1) %
              markdownVersions.length;
          _updateText(nextIndex);
        },
      ),
    );
  }
}

void main() => runApp(const MaterialApp(
      home: FadeRevealMarkdownDifference(),
    ));
```
> End of Code 
<br>
Feel free to let me know if you need any modifications or additional customizations. I’m open to making adjustments to better fit your needs.

###  Something off-topic
### Need  Feedback  for this too , may be you Like that
Out my two Ideas I tried. I liked my other idea which is Typewriter kind of Stuff , May be You can Find it helpful , I am leaving a code snippet for that too , Will wait for you feedback
> This is extra Stuff 

```dart
import 'package:flutter/material.dart';
import 'package:flutter_markdown/flutter_markdown.dart';

class TypewriterMarkdownDifference extends StatefulWidget {
  @override
  _TypewriterMarkdownDifferenceState createState() =>
      _TypewriterMarkdownDifferenceState();
}

class _TypewriterMarkdownDifferenceState
    extends State<TypewriterMarkdownDifference>
    with SingleTickerProviderStateMixin {
  final List<String> markdownVersions = [
    "## Problem\nThis is the main issue with what",
    "## Problem\nThis is the main issue with what we're trying to solve.",
    "## Problem\nThis is the main issue with what we're trying to solve. The current solution is insufficient for handling edge cases.",
    "## Problem\nThis is the main issue with what we're trying to solve. The current solution is insufficient for handling edge cases, and it leads to frequent breakdowns in production.",
    "## Problem\nThis is the main issue with what we're trying to solve. The current solution is insufficient for handling edge cases, and it leads to frequent breakdowns in production. As a result, customer satisfaction has dropped significantly.",
  ];

  late AnimationController _controller;
  late Animation<int> _characterCountAnimation;

  String previousText = "";
  String currentText = "";
  String newText = "";

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
      duration: const Duration(seconds: 2), // Speed of typewriter effect
      vsync: this,
    );

    // Initialize text and animation for the first version
    _updateText(0);
  }

  void _updateText(int versionIndex) {
    if (versionIndex < 0 || versionIndex >= markdownVersions.length) return;

    setState(() {
      previousText = versionIndex > 0 ? markdownVersions[versionIndex - 1] : "";
      currentText = markdownVersions[versionIndex];
      newText = _findNewText(previousText, currentText);

      // Reinitialize animation based on the new text length
      _characterCountAnimation = StepTween(
        begin: 0,
        end: newText.length,
      ).animate(
        CurvedAnimation(
          parent: _controller,
          curve: Curves.easeInOut,
        ),
      );

      // Reset and start the animation
      _controller.reset();
      _controller.forward();
    });
  }

  // Compare previous and current text to return the newly added portion
  String _findNewText(String oldText, String newText) {
    if (oldText == newText) return "";
    return newText.substring(oldText.length); // Get new text after old text
  }

  @override
  void dispose() {
    _controller.dispose(); // Dispose of the controller to free resources
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Typewriter Markdown Difference Effect"),
      ),
      body: Padding(
        padding: const EdgeInsets.all(16.0),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Expanded(
              child: AnimatedBuilder(
                animation: _characterCountAnimation,
                builder: (context, child) {
                  // Extract the part of the new text that's visible based on animation
                  String visibleText = newText.isNotEmpty
                      ? newText.substring(0, _characterCountAnimation.value)
                      : "";

                  // Combine the old static markdown and the animated new markdown
                  String fullMarkdown = previousText + visibleText;

                  return MarkdownBody(
                    data: fullMarkdown,
                    styleSheet: MarkdownStyleSheet(
                      h2: const TextStyle(fontSize: 24, fontWeight: FontWeight.bold),
                      p: const TextStyle(fontSize: 18, color: Colors.black87),
                    ),
                  );
                },
              ),
            ),
            ElevatedButton(
              onPressed: () {
                // Move to the next version and restart the animation
                int nextIndex = (markdownVersions.indexOf(currentText) + 1) % markdownVersions.length;
                _updateText(nextIndex);
              },
              child: const Text("Next Version"),
            ),
          ],
        ),
      ),
    );
  }
}

void main() => runApp(MaterialApp(
      home: TypewriterMarkdownDifference(),
    ));
``` 


https://github.com/user-attachments/assets/1e2e7112-81ab-4812-b541-c9b236c483d8

FilledStacks commented:
Hey, 

Your solution is the closest, but the problem still exists where the text that was already animated in lightly pulses when the new text is added.

You can see it in the preview.

mg3994 committed:
TweenAnimationBuilder Used and Value key for Widget Animation

I've noticed that it can be challenging to accurately identify whether a visual effect is flickering or pulsing when wearing spectacles. The lenses can introduce optical distortions that can make it difficult to discern subtle differences in light intensity or frequency. So, I increased Duration

mg3994 commented:
Hey @FilledStacks ,
You have excellent eyesight! I'm impressed with your ability to catch even the smallest details.

I appreciate your feedback and the time you've taken to provide a more detailed explanation of the issue. It seems the problem persists even with the proposed solution, where the previously animated text experiences slight pulsing when new text is introduced.

I'll be sure to take this into account as I continue to explore potential solutions. Perhaps there's a way to optimize the animation timing or transitions to prevent this interference.

mg3994 commented:
New PR SOON with `extensionSet:`

FilledStacks commented:
@mg3994 no worries I'll accept this solution as it's the closest to what I asked for. 

I also like the implementation you used. 

thank you. 

You're the winner of this bounty 🚀

FilledStacks commented:
Please leave me an email here to contact you on

mg3994 commented:
I'm honored to have won your bounty. Your trust and recognition mean a lot to me. I'll continue to learn and grow as a developer, striving to deliver exceptional results. Thank you for the opportunity. I understand that the solution may not have met your exact pixel-perfect expectations. I apologize for any shortcomings and appreciate your understanding. I'll continue to strive for perfection in future projects. 
**Below is my Email:**
###### [Email linked with receiving newsletter from Ticket Stacker]
```bash
manishgautammg7@gmail.com
```
Personal Email:
```bash
manishsharma3994@gmail.com
```

FilledStacks commented:
Awesome, I'm very happy to hear that. 

I have emailed you, let's get this bounty settled 😁

mg3994 commented:
**Thanks a Lot**
<br>
![Screenshot 2024-11-04 194253](https://github.com/user-attachments/assets/662253f1-61fb-4b33-b9b4-656bb809cb1a)
